src/KPM.jl

```diff
@@ -24,9 +24,9 @@ function KPMBuilder(h, A, kets, order, bandrange)
     eA = eltype(eltype(A))
     mulist = zeros(promote_type(eh, eA), order + 1)
     bandbracket = bandbracketKPM(h, bandrange)
-    h´ = matrixKPM(h)
-    A´ = matrixKPM(A)
-    kets´ = Matrix(kets, h)
+    h´ = matrixKPM(h, SparseMatrixCSC{eh, Int64})
+    A´ = matrixKPM(A, SparseMatrixCSC{eh, Int64})
+    kets´ = matrixKPM(kets, h)
```

Comment:
Nice, this is the key of the fix. We need a `matrixKPM(kets, h)` that works with whatever type `kets`, flattening `kets` if its eltype is not a `Number`. 

So we need a `matrixKPM(kets::Matrix{<:Number},h) = kets` that is a no-op, like you've done below. And then we need a fallback `matrixKPM(kets,h)` that flattens kets. Instead of writing a specific method, we should simply reuse all the machinery we have for flatten. This is important, because there is substantial complexity in the case of different sublattices with different number of orbitals that need to be dealt with correctly. Fortunately we already have methods for that, specifically `flatten_dense_copy!` and `flatten_sparse_copy!`

Since you want a dense flat matrix as output (because sparse*dense is what we want for speed), you need to call `flatten_dense_copy!(dst, src, o::OrbitalStructure)`. Now, the `OrbitalStructure` is that of `h`. So `matrixKPM(kets, h)` should (1) check that the eltype of `kets` is the same as that of the unflattened `h`, and (2) that the length is also correct. Then it must extract `o = orbitalstructure(h)` and prepare a `s = similarmatrix(ket, flatten)`. Finally return a `flatten_dense_copy!(s, kets, o)`.

All this is just a sketch of course.

Reply:
Yes! I agree. It makes sense to use Quantica internal functions instead of defining new routines. However, in order to do so, we need to generalise ` flatten_dense_copy!` so it accepts non square AbstractMatrices, this functionality is not already built-in